Meadowlark-js/backends/meadowlark-postgresql-backend/src/repository/Upsert.ts

```diff
@@ -4,26 +4,43 @@
 // See the LICENSE and NOTICES files in the project root for more information.
 
 import type { PoolClient, QueryResult } from 'pg';
-import { UpsertResult, UpsertRequest, Logger } from '@edfi/meadowlark-core';
-import { getDocumentInsertOrUpdateSql, getRecordExistsSql } from './QueryHelper';
+import {
+  UpsertResult,
+  UpsertRequest,
+  Logger,
+  DocumentReference,
+  documentIdForDocumentReference,
+} from '@edfi/meadowlark-core';
+
+import {
+  getDeleteReferencesSql,
+  getDocumentInsertOrUpdateSql,
+  getCheckDocumentExistsSql,
+  getReferencesInsertSql,
+} from './QueryHelper';
+import { validateReferences } from './WriteHelper';
 
 export async function upsertDocument(
   { id, resourceInfo, documentInfo, edfiDoc, validate, traceId, security }: UpsertRequest,
   client: PoolClient,
 ): Promise<UpsertResult> {
-  const upsertResult: UpsertResult = { response: 'UNKNOWN_FAILURE' };
+  let upsertResult: UpsertResult = { response: 'UNKNOWN_FAILURE' };
 
   let recordExistsResult: QueryResult;
-  // let outRefs;
-  let documentSql: string;
+  let documentUpsertSql: string;
   let isInsert: boolean;
 
+  const outRefs = documentInfo.documentReferences.map((dr: DocumentReference) => documentIdForDocumentReference(dr));
+
   try {
-    recordExistsResult = await client.query(getRecordExistsSql(id));
+    recordExistsResult = await client.query(getCheckDocumentExistsSql(id));
```

Comment:
I don't think this works. Your existence check here happens outside the transaction.

More broadly though, PostgreSQL has an upsert syntax that would simplify this. Here's an example of the syntax: https://towardsdatascience.com/postgresql-how-to-upsert-safely-easily-and-fast-246040514933

Reply:
Say the document already exists. This query returns true, and you are ready for an update. Then a different request deletes the document before you enter your transaction. In your transaction, you choose the "UPDATE" SQL, and error because you are updating something that isn't there.

This does bring up another issue not handled here. Say we bring the check into the transaction (we should), but otherwise assume the same sequence. The transaction itself will fail, which is fine but it needs to be retried (with some sort of backoff). The MongoDB driver handles retries for you in this situation, but I'm assuming we'll need to handle it ourselves here.

Reply:
I'm not sure why I originally had the check outside of the transaction, but I went ahead and moved it into the transaction (and a single try/catch)